### ISSUE TITLE
Regressions in the navigation UI

### ISSUE THREAD

**Issue by ErisDS** [TryGhost/Ghost#5841]
Having read a few bug reports, and played around with the Navigation UI a bit since 0.7, it would appear there are a few regressions in how it works.

A few issues I've noticed:

### 1. Cmd & S doesn't work unless you focus-out

- create a new item, with a label and a url
- go back and change the url, whilst still in the field - hit cmd+s 
- your changes disappear

### 2. URLs are being stored absolute rather than relative

- add a link for a pretty standard url like 'about'
- the UI should show you the full absolute URL
- the database should contain the relative part only, and does not

I think this is the underlying cause of https://github.com/TryGhost/Ghost/issues/5801

### 3. Validation has gone wonky

- Validation for navigation has been too strict from day one: #4989
- There were some fixes to allow mailto: #5581 #5013
- I tried adding my email address as a mailto link when testing that that now works, and forgot to add `mailto:` and discovered that email addressed are accepted as valid urls, which seems odd.

### 4. `{{url}}` helper not working to output urls correctly in the `@blog.navigation` context 

- This was reported here: https://github.com/TryGhost/Ghost/issues/5162#issuecomment-138122876 and here https://github.com/TryGhost/Ghost/issues/5836#issuecomment-140156842
- This is probably not related to the UI in any way, I'm just dumping it here so I don't lose track

#### Things to fix / improve

- [x] cmd + s saves correctly
- [x] urls are stored relatively
- [x] mailto links are still allowed
- [x] `{{current}}` helper works again
- [ ] `{{url}}` helper works whenever there's a navigation item around
- [ ]  *bonus* validation is improved so it permits valid URLs, and doesn't allow emails
- [ ]  *bonus* get this working with HTTP vs HTTPS (e.g. permit urls with no protocol)

There are many ways the UX of this could be improved, as well as the implementation (by doing more work server side) suggestions are welcome :)

**Comment by mwakerman** [comment]
Also relevant #5839 
@kevinansfield @ErisDS - I was having an issue where my blog is served both secure and unsecure but I want to be able to choose if a nav link sends a user to a https or a http endpoint. There was a bug here which I think I've fixed but it would also not be possible if the DB only stored the relative path.

**Comment by kevinansfield** [comment]
@mwakerman I'm curious as to your requirement, could you explain the use case please?

With the changes in #5839, any URLs where the host matches will be stored as relative so that HTTP/HTTPS is determined by the current protocol. This makes sense to me because if a page should be served via HTTPS then that should be enforced higher up the chain with a redirect because simply relying on links to switch to the correct protocol would be insecure.

**Comment by mwakerman** [comment]
@kevinansfield, I agree, the link should use the same protocol as the current page. My fear (should've thought a bit more) was that storing the relative would take people to the unsecured endpoint even from a secured page.

**Comment by kevinansfield** [comment]
@ErisDS #5843 should be good to go once tests have confirmed my wonky rebase hasn't broken anything. I've limited it to only Ember changes as it turned into a pretty epic PR and thought the server bits may be best kept separate.

It covers everything in the list apart from the `{{url}}` and `{{current}}` helpers and probably the HTTP vs HTTPS item as I'm assuming that's a server-side issue?

Not sure if this was implied with the `{{url}} / {{current}}` todo items but on the front-end any alternative scheme, protocol-relative or plain anchor URLs (`tel:`, `javascript:`, `//example.com`, `#contact`, etc) get the blog URL prepended breaking their functionality.

**Comment by ErisDS** [comment]
A large majority of the bugs with `{{url}}` and `{{current}}` and also HTTP/HTTPS were caused by the issue with absolute URLs being stored instead of relative. As soon as we're back to relative URLs then the helpers can figure out what to do next :) Any remaining issues I'll try to patch once this is merged in.

The "alternative scheme, protocol-relative or plain anchor URLs" working properly was roughly covered by the bonus item: "validation is improved so it permits valid URLs, and doesn't allow email", but this can be fixed later against #4989.

**Comment by kevinansfield** [comment]
Those broken URLs pass validation and are handled properly in the admin, it's in the `{{navigation}}` helper where they break. I made a change in `server/config/url.js` that fixes it but I'm not sure what other ramifications that will have, I'll push as a new PR.

**Comment by ErisDS** [comment]
Oh that's great - as long as it's all fixed in the UI that's fantastic. If you've got a PR for the url functions that are causing the other side of the problem then cool throw it up :)

**Comment by ErisDS** [comment]
Notes from the PR that was merged:

> Save & validation seem to be working much better. The email handling doesn't seem to be quite right just yet - it adds mailto when you press enter or focus out, but then adds the full URL again on save - not sure if that's because #5864 is needed?

> Also noticed we've lost the behaviour which added the trailing slash at the end of a relative URL and can't remember if that was done on purpose to fix another issue or not. I don't think so. Without adding the slash to genuine relative URLs, the current helper will probably still not work - but might be something to fix separately to this already pretty epic PR?

I think the current helper probably needs to be addressed slightly differently - I've been looking at how to make some changes to how the theme side of things work so that this is more consistent as part of fixing #5162.

**Comment by ErisDS** [comment]
Closing for now, will open specific issues for any remaining work
